
<!DOCTYPE html>
<html lang="en">
    
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="generator" content="VictoryWoo">
    <title>[Java] 14. 예외 처리 - VictoryWoo</title>
    <meta name="author" content="VictoryWoo">
    
    
    
    <script type="application/ld+json">{"@context":"http://schema.org","@type":"BlogPosting","author":{"@type":"Person","name":"VictoryWoo","sameAs":["https://github.com/WooVictory","https://www.facebook.com/WooVictory","https://www.linkedin.com/profile/","jhsw0375@gmail.com"],"image":"Victory.jpeg"},"articleBody":"오류(error)와 예외(exception)\n\n자바 프로그램을 작성할 때 자바 문법에 맞지 않게 코드를 작성하고 컴파일하려고 하면 자바 컴파일러는 문법 오류(syntax error)를 발생시킵니다. 또한, 자바 문법에는 맞게 작성되었다 하더라도 프로그램이 실행되면서 예상하지 못한 오류가 발생할 수 있습니다.\n이렇게 컴퓨터 시스템이 동작하는 도중에 예상하지 못한 사태가 발생하여 실행 중인 프로그램이 영향을 받는 것을 오류(error)와 예외(exception) 두 가지로 구분할 수 있습니다.\n\n코드를 작성하는 과정에서 예기치 못한 수많은 에러들이 발생합니다. 예를 들면 다음과 같습니다.\n컴파일 전에 알게 되는 에러도 있고, 작성할 때는 아무 문제 없다가 실행시키고 나서야 발생하는 에러들이 있습니다. 이것들을 제어할 수 있어야 좀 더 안전한 프로그램을 설계할 수 있습니다.\n\n에러(error)? 예외(exception)?\n컴파일 에러 - 컴파일 시에 발생하는 에러\n런타임 에러 - 실행 시에 발생하는 에러\n논리적 에러 - 실행은 되지만, 의도와 다르게 동작하는 에러\n그럼 위에서 말하는 컴파일, 런타임은 무엇이면 컴파일 에러는 무엇일까요?? 공부를 하다보니 들어본 적은 있지만, 정확하게 무엇이라고 설명하기에는 부족한 부분들이 있어서 아래에 정리해봤습니다.\n\n컴파일 에러 : 주로 문법상의 오류가 많습니다. (빨간 줄이 나타나는 에러와 같은 경우입니다. 대부분의 IDE들이 빨간줄 표시로 잡아줍니다.) 컴파일 에러가 발생하는 경우는 주로 오타, 중괄호의 짝이 맞지 않는 경우, 중복된 변수명 등의 원인이 있을 수 있습니다.\nRunTime : 어떤 프로그램이 실행되는 동안의 Time을 의미하며, runtime error는 어떤 프로그램이 실행되는 동안 발생하는 에러입니다.\n런타임 에러 : 문법상으로는 에러가 없지만 실행시켰을 때 에러가 나는 경우, 대표적인 경우가 NullpointerException\nCompile Time : 고급 언어(프로그래밍 언어)를 기계어로 변경하는 과정이라고 하며, compile error는 고급 언어(프로그래밍 언어)를 기계어로 변경하는 과정에서 발생하는 에러입니다.\n\n즉, 컴파일 에러가 발생하지 않더라도 런타임 에러가 발생할 수 있습니다.\n\n오류(error)\n**오류(error)**는 시스템 레벨에서 프로그램에 심각한 문제를 야기하여 실행 중인 프로그램을 종료시킵니다. 이러한 오류는 개발자가 미리 예측하여 처리할 수 없는 것이 대부분이므로, 오류에 대한 처리는 할 수 없습니다.\n(ex. 메모리 부족, 스택 오버플로우와 같이 발생하면 복구할 수 없는.)\n예외(exception)\n**예외(exception)**는 오류와 마찬가지로 실행 중인 프로그램을 비정상적으로 종료시키지만, 발생할 수 있는 상황을 미리 예측하여 처리할 수 있습니다. 따라서 개발자는 예외 처리(Exception handling)를 통해 예외 상황을 처리할 수 있도록 코드의 흐름을 바꿀 필요가 있습니다.\n(발생하더라도 수습될 수 있는, 비교적 덜 심각한. 그렇기에 프로그램의 비정상적인 종료를 사전에 예방할 수 있음.)\n예외처리(exception handling)\n\n자바에서 예외 처리는 이 Exception을 Handling하는 것을 말합니다. 프로그램을 제작하는 과정에서 발생할 예외에 대해 미리 코드를 작성하는 것입니다.\nif-else문으로 처리를 해야만 했던, C에서는 예외 처리를 위한 if문과 일반 if문을 구분하기 힘들었습니다. 이 문제를 해결하기 위해 Java에서는 예외 처리에 특화된 문법을 제공합니다.\n그것이 바로 try-catch문입니다. 사용하는 방법과 코드에 대한 설명은 아래와 같습니다.\n1234567891011121314try&#123;// 예외를 발생시킬 위험을 갖고 있는 코드(로직)// 이 로직은 예외 상황과 관련있는 문장들도 고려해야 합니다.&#125;catch(Exception e1)&#123;// 예외가 발생했을 때 실행되는 로직// Exception클래스는 모든 예외의 상위 클래스&#125;catch(e2)&#123;&#125;...finally&#123;// 예외 발생 여부와 상관없이 무조건 실행될 코드&#125;\n위의 코드에서 보면 예외를 발생시킬 위험을 갖고 있는 코드인 경우 try{ } 블록 안에 넣어둡니다. 그리고 catch를 통해 오류 메시지를 출력합니다. try안에 있는 코드를 실행하다가 오류가 발생하면 catch 구문으로 Jump합니다.\n\nJump : 오류가 발생한 구문 다음의 코드는 실행되지 않는다.\n\n그러면 catch 메소드가 실행되게 되는데, 이 때 Exception이라는 클래스를 e로 인스턴스화하여 catch에 매개변수로 전달합니다. 그리고 Exception 이라는 클래스에 있던 getMessage라는 메소드를 호출하거나 다른 구문을 실행합니다.\n다음은 예외가 발생했을 때 실행되는 로직의 예입니다.\nSysetm.out.println(e.getMessage()); -&gt; 가장 간단한 예외 상황을 출력\nSysetm.out.println(e.toString()); -&gt; 예외 상황에 대한 좀 더 자세한 정보를 출력\ne.printStackTrace(); -&gt; 예외 상황이 발생한 소스코드의 위치까지 출력\n그리고 위의 코드처럼 catch 구문을 여러 개 두어서 다중 catch도 가능합니다.\ncatch 블록과 finally 블록은 선택적인 옵션으로 반드시 사용할 필요는 없습니다. 따라서 사용할 수 있는 모든 적합한 try 구문은 다음과 같습니다.\n1231. try / catch2. try / finally3. try / catch / ... / finally\n\n다른 제어문과 달리 예외 처리문은 중괄호를 생략할 수 없습니다.\n\n예외처리 메커니즘\n자바에서 예외 처리는 아래의 그림과 순서로 진행됩니다.\n\ntry 블록에 도달한 프로그램의 제어는 try 블록 내의 코드를 실행합니다.\n이 때 만약 예외가 발생(throw)하지 않고, finally 블록이 존재하면 프로그램의 제어는 바로 finally 블록으로 이동합니다.\ntry 블로에서 예외가 발생하면 catch 구문으로 Jump하게 되고, catch 핸들러는 다음과 같은 순서로 적절한 catch 블록을 찾게 됩니다.\n\n스택에서 try 블록과 가장 가까운 catch 블록부터 차례대로 검사합니다.\n만약 적절한 catch 블록을 찾지 못하면, 바로 다음 바깥쪽 try 블록 다음에 위치한 catch 블록을 차례대로 검사합니다.\n이러한 과정을 가장 바깥쪽의 try 블록까지 계속 검사하게 됩니다.\n그래도 적절한 catch 블록을 찾지 못하면, 예외는 처리되지 못합니다.\n\n\n만약 적절한 catch 블록을 찾게 되면, throw 문의 피연산자는 예외 객체의 형식 매개변수로 전달됩니다.\n모든 예외 처리가 끝나면 프로그램의 제어는 finally 블록으로 이동합니다.(finally 블록이 있다면)\nfinally 블록이 모두 처리되면, 프로그램의 제어는 예외 처리문 바로 다음으로 이동합니다.\n\n\n만약 1번 try 블록에서 예외가 발생하지 않고, 바깥쪽 try 블록에서도 예외가 발생하지 않으면, 6번 finally 블록이 바로 실행될 것입니다.\n하지만, 1번 try 블록에서 예외가 발생하면 2번과 3번 catch 블록에서 해당 예외를 처리할 수 있는지 검사하게 됩니다. 만약 적절한 catch 블록을 처리하지 못하면, 바깥쪽 4,5번 catch 블록도 차례대로 검사하게 됩니다.\n이 때 해당 예외를 처리할 수 있는 catch 블록을 찾게 되면, 해당 catch 블록을 실행한 후 6번의 finally 블록을 실행합니다. 하지만 모든 catch 블록이 해당 예외를 처리할 수 없으면, 예외는 처리되지 못한채 해당 프로그램을 강제 종료될 것입니다.\nException 클래스\n\n자바에서 모든 예외의 조상 클래스가 되는 Exception 클래스는 크게 다음과 같이 구분할 수 있습니다.\n\nRuntimeException 클래스\n그 외의 Exception 클래스의 자식 클래스\n\n\nRuntimeException 클래스를 상속받는 자식 클래스들은 주로 치명적인 예외 상황을 발생시키지 않는 예외들로 구성됩니다. 따라서 try / catch 문을 사용하기보다는 프로그램을 작성하면서 예외가 발생하지 않도록 주의를 기울이는 편이 좋습니다.\n하지만 그 외의 Exception 클래스에 속하는 자식 클래스들은 치명적인 예외 상황을 발생시키므로, 반드시 try / catch문을 사용하여 예외를 처리해야만 합니다. 따라서 자바 컴파일러는 RuntimeException 클래스 이외의 Exception 클래스의 자식 클래스에 속하는 예외가 발생할 가능성이 있는 구문에는 반드시 예외를 처리하도록 강제하고 있습니다. 만약 이러한 예외가 발생할 가능성이 있는 구문을 처리하지 않았을 때는 컴파일 시 오류를 발생시킵니다.\n다음 예제는 PrintStream 클래스의 write() 메소드를 사용하여 byte 타입 배열의 모든 요소를 출력하는 예제입니다. 하지만 write() 메소드에서 발생할 수 있는 IOException에 대한 예외를 처리하지 않았으므로, 컴파일 시 오류가 발생합니다.\n따라서 try / catch문을 사용하여 IOException에 대한 예외 처리까지 해주어야만 컴파일 할 수 있습니다.\n12345678import java.io.IOException;public class ExceptionExam &#123;    public static void main(String[] args) throws IOException &#123;        byte[] list = &#123;1,3,2&#125;;        System.out.write(list);    &#125;&#125;\n예외 처리의 계층 관계\n자바에서는 예외가 발생하면, try 블록과 가장 가까운 catch 블록부터 순서대로 검사합니다. 따라서 여러 개의 catch 블록을 사용할 때는 Exception 클래스의 계층 관계에도 주의를 기울여야만 합니다.\n12345678910111213try &#123;    System.out.write(list);&#125; catch (Exception e) &#123;    e.printStackTrace();&#125; catch (IOException e) &#123;    e.printStackTrace();&#125;\n위의 예제에서 IOException이 발생하면, 자바는 첫 번째 catch 블록부터 순서대로 해당 예외를 처리할 수 있는지를 검사합니다. 그런데 IOException은 Exception 클래스의 자식 클래스이므로, 첫 번째 catch 블록에서도 IOException을 처리할 수 있습니다.\n따라서 IOException을 비롯한 Exception 클래스의 자식 클래스에 해당하는 예외가 발생하면, 언제나 첫 번째 catch 블록에서만 처리될 것입니다.\n즉, catch 블록의 순서를 위의 코드처럼 작성하면, 두 번째 catch 블록은 영원히 실행되지 않을 것입니다.\n따라서, IOException만을 따로 처리하고자 한다면, 다음과 같이 catch 블록의 순서를 변경해야 합니다.\n1234567891011121314try &#123;    System.out.write(list);&#125; catch (IOException e) &#123;    e.printStackTrace();&#125;catch (Exception e) &#123;    e.printStackTrace();&#125;\n이렇게 변경하면 IOException이 발생하면, 첫 번째 catch 블록에서 해당 예외를 처리할 것입니다. 또한, IOException 이외의 Exception 클래스의 자식 클래스에 해당하는 예외가 발생하면 두 번째 catch 블록에서 처리될 것입니다. 이처럼 범위가 더 좁은 예외를 처리하는 catch 블록을 먼저 명시하고, 범위가 더 넓은 예외를 처리하는 catch 블록은 나중에 명시해야만 정상적으로 해당 예외를 처리할 수 있습니다.\n여러 예외 타입의 동시 처리\nJava SE 7 부터는 ‘|’ 기호를 사용하여 하나의 catch 블록에서 여러 타입의 예외를 동시에 처리할 수 있습니다.\n12345678910111213try &#123;    this.db.commit();&#125; catch (IOException e) &#123;    e.printStackTrace();&#125; catch (SQLException e) &#123;    e.printStackTrace();&#125;\n위와 같이 일반적으로 작성하던 catch 블록을 아래와 같이 동시에 처리할 수 있습니다.\n123456789try &#123;    this.db.commit();&#125; catch (IOException | SQLException e) &#123;    e.printStackTrace();&#125;\n하지만, 둘 이상의 예외 타입을 동시에 처리하는 catch 블록에서 매개변수로 전달받은 예외 객체는 묵시적으로 final 제어자를 가지게 됩니다. 따라서 catch 블록 내에서 해당 매개변수에는 어떠한 값도 대입할 수 없습니다.(상수화가 되었기 때문)\nThrowable 클래스\n자바에서 Throwable 클래스는 모든 예외의 조상이 되는 Exception 클래스와 모든 오류의 조상이 되는 Error 클래스의 부모 클래스입니다. Throwable 타입과 이 클래스를 상속받은 서브 타입만이 자바 가상 머신(JVM)이나 throw 키워드에 의해 던져질 수 있습니다.\n이 클래스에는 예외나 오류에 관한 다양한 정보를 확인할 수 있는 다음과 같은 메소드가 포함되어 있습니다.\n\nString getMessage() : 해당 throwable 객체에 대한 자세한 내용을 문자열로 반환함\nvodi printStackTree() : 해당 throwable 객체와 표준 오류 스트림(standard error stream)에서 해당 객체의 스택 트레이스를 출력함\nString toString() : 해당 throwable 객체에 대한 간략한 내용을 문자열로 반환함\n\n다음 예제는 일부로 숫자를 0으로 나눠서 ArithmeticException 오류를 발생시키는 예제입니다. 이렇게 발생한 오류에 대해서 Throwable 메소드를 사용하여 발생한 오류에 대한 정보를 출력합니다.\n12345678910111213141516package com.company;import java.io.IOException;public class ExceptionExam &#123;    public static void main(String[] args)&#123;        try&#123;            System.out.println(5/0);        &#125;catch (ArithmeticException e)&#123;            e.printStackTrace();        &#125;    &#125;&#125;// 결과현재 발생한 예외의 정보 : / by zero\n자주 사용되는 예외 클래스\n자바에서 자주 사용되는 예외 클래스는 다음의 표와 같습니다.\n\n예외 발생 및 회피\n\n자바에서는 throw 키워드를 사용하여 강제로 예외를 발생시킬 수 있습니다.\n123Exception e = new Exeption(\"메시지\");throw e;\n예외 회피하기\n예외 처리에 대한 부분에서 가장 헷갈리는 부분이 throw와 throws입니다.\n\nthrow\n\n예외를 발생시키는 것이고, 예외를 던지는 것입니다.\n메소드 내에서 상위 블록으로 예외를 던지는 것\n억지로 에러를 발생시킬 때도 사용되지만 현재 메소드의 에러를 처리한 후에 상위 메소드에 에러 정보를 줌으로써 상위 메소드에서도 에러가 발생한 것을 감지할 수 있습니다.\n실제로 exception을 throw할 때 사용하는 키워드\nthrow는 강제로 예외를 발생시키는 것\n\n\n\n123456789101112131415161718public class ExceptionExam &#123;    static void handlingException() &#123;        try &#123;            throw new Exception();        &#125; catch (Exception e) &#123;            System.out.println(\"호출된 메소드에서 예외가 처리됨!\");        &#125;    &#125;    public static void main(String[] args) &#123;        try &#123;            handlingException();        &#125; catch (Exception e) &#123;            System.out.println(\"main() 메소드에서 예외가 처리됨!\");        &#125;    &#125;&#125;// 결과호출된 메소드에서 예외가 처리됨!\nMain 함수에서 handlingException() 함수를 호출하고 그 함수 안에서는 throw 키워드를 사용하여 예외를 발생시키고 있습니다. 이 예외를 발생시키는 문장을 try / catch로 감싸고 있기 때문에 발생한 예외는 catch 블록으로 Jump하게 되고 함수 내에서 예외를 처리하게 됩니다.\n\nthrows\n\n메소드나 생성자를 수행할 때 발생하는 Exception을 선언할 때 사용하는 키워드\n예외를 던지는 것\n현재 메소드에서 상위 메소드로 예외를 던집니다.\n다시 말해, 에외를 자신이 직접 처리하지 않고, 자신을 호출한 메소드에세 책임을 전가하는 것입니다.\n\n\n\n123456789101112131415public class Exception04 &#123;    static void handlingException() throws Exception &#123;         throw new Exception();     &#125;    public static void main(String[] args) &#123;        try &#123;            handlingException();        &#125; catch (Exception e) &#123;            System.out.println(\"main() 메소드에서 예외가 처리됨!\");        &#125;    &#125;&#125;// 결과main() 메소드에서 예외가 처리됨!\n위의 코드는 throws를 사용하여 handlingException() 함수에서 발생한 예외를 전가시키는 예제입니다. 메소드를 정의할 때 throws 키워드를 추가하여 이 메소드를 호출하는 곳에서 예외 처리를 하도록 전가시킬 수 있습니다.\nhandlingException() throws Exception 라는 문장을 해석해보자면, handlingException()라는 함수가 Exception 예외를 던진다는 뜻이므로 handlingException() 함수를 사용하는 곳(혹은 호출하는 곳)으로 예외를 던집니다.\n여기서는 Main 함수에서 handlingException() 함수를 호출하였고 handlingException() 함수는 이 함수를 호출한 곳으로 예외를 던지기 때문에 Main 함수에서 예외를 처리해주어야 합니다. 하지만, 예외가 발생할 수 있기 때문에 앞에서 배운 에외 처리를 통해서 함수를 호출하는 문장을 try 블록으로 감싸줍니다. 그러면 여기서 예외가 발생하면 아래의 catch 블록으로 빠지게 되고, 결국에는 Main 함수에서 예외를 처리하게 됩니다.\n사용자 정의 예외 클래스\n자바에서는 Exception 클래스를 상속받아 자신만의 새로운 예외 클래스를 정의하여 사용할 수 있습니다.\n사용자 정의 예외 클래스에는 생성자뿐 아니라 필드 및 메소드도 원하는 만큼 추가할 수 있습니다.\n12345class MyException extends RuntimeException&#123;    MyException(String errMsg)&#123;        super(errMsg);    &#125;&#125;\n요즘에는 위와 같이 Exception 클래스가 아닌 예외 처리를 강제하지 않는 RuntimeException 클래스를 상속받아 작성하는 경우가 많습니다.\ntry-with-resources문\nJava SE 7 부터는 사용자 자원을 자동으로 해제해주는 try-with-resources 문을 사용할 수 있습니다.\n123try(파일을 열거나 자원을 할당하는 명령문)&#123;    ...&#125;\n위와 같이 try 블록에 괄호 () 를 추가하여 파일을 열거나 자원을 할당하늠 명령문을 명시하면, 해당 try 블록이 끝나자마자 자동으로 파일을 닫거나 할당된 자원을 해제해줍니다.\n다음은 파일에서 문자열을 한 줄 읽어오는 예제입니다.\n123456789static String readFile(String filePath) throws IOException &#123;    BufferedReader br = new BufferedReader(new FileReader(filePath));    try &#123;        return br.readLine();    &#125; finally &#123;        if (br != null)            br.close();    &#125;&#125;\n위와 같이 Java SE 7 이전에서는 finally 블록을 사용하여 사용한 파일을 닫아줘야 했습니다. 하지만 try-with-resources 문을 사용하면 다음과 같이 자동으로 파일의 닫기를 수행할 수 있습니다.\n12345static String readFile(String filePath) throws IOException &#123;    try (BufferedReader br = new BufferedReader(new FileReader(filePath))) &#123;        return br.readLine();    &#125;&#125;\n","dateCreated":"2018-09-02T11:02:54+09:00","dateModified":"2020-04-22T16:47:55+09:00","datePublished":"2018-09-02T11:02:54+09:00","description":"","headline":"[Java] 14. 예외 처리","image":[],"mainEntityOfPage":{"@type":"WebPage","@id":"https://woovictory.github.io/2018/09/02/java-basic-14/"},"publisher":{"@type":"Organization","name":"VictoryWoo","sameAs":["https://github.com/WooVictory","https://www.facebook.com/WooVictory","https://www.linkedin.com/profile/","jhsw0375@gmail.com"],"image":"Victory.jpeg","logo":{"@type":"ImageObject","url":"Victory.jpeg"}},"url":"https://woovictory.github.io/2018/09/02/java-basic-14/","keywords":"Java"}</script>
    <meta name="description" content="오류(error)와 예외(exception)  자바 프로그램을 작성할 때 자바 문법에 맞지 않게 코드를 작성하고 컴파일하려고 하면 자바 컴파일러는 문법 오류(syntax error)를 발생시킵니다. 또한, 자바 문법에는 맞게 작성되었다 하더라도 프로그램이 실행되면서 예상하지 못한 오류가 발생할 수 있습니다. 이렇게 컴퓨터 시스템이 동작하는 도중에 예상하지">
<meta name="keywords" content="Java">
<meta property="og:type" content="blog">
<meta property="og:title" content="[Java] 14. 예외 처리">
<meta property="og:url" content="https://woovictory.github.io/2018/09/02/java-basic-14/index.html">
<meta property="og:site_name" content="VictoryWoo">
<meta property="og:description" content="오류(error)와 예외(exception)  자바 프로그램을 작성할 때 자바 문법에 맞지 않게 코드를 작성하고 컴파일하려고 하면 자바 컴파일러는 문법 오류(syntax error)를 발생시킵니다. 또한, 자바 문법에는 맞게 작성되었다 하더라도 프로그램이 실행되면서 예상하지 못한 오류가 발생할 수 있습니다. 이렇게 컴퓨터 시스템이 동작하는 도중에 예상하지">
<meta property="og:locale" content="en">
<meta property="og:image" content="https://woovictory.github.io/img/compile_runtime.png">
<meta property="og:image" content="https://woovictory.github.io/img/try_catch.png">
<meta property="og:image" content="https://woovictory.github.io/img/exception.png">
<meta property="og:image" content="https://woovictory.github.io/img/exception_class.png">
<meta property="og:updated_time" content="2020-04-22T07:47:55.906Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="[Java] 14. 예외 처리">
<meta name="twitter:description" content="오류(error)와 예외(exception)  자바 프로그램을 작성할 때 자바 문법에 맞지 않게 코드를 작성하고 컴파일하려고 하면 자바 컴파일러는 문법 오류(syntax error)를 발생시킵니다. 또한, 자바 문법에는 맞게 작성되었다 하더라도 프로그램이 실행되면서 예상하지 못한 오류가 발생할 수 있습니다. 이렇게 컴퓨터 시스템이 동작하는 도중에 예상하지">
<meta name="twitter:image" content="https://woovictory.github.io/img/compile_runtime.png">
    
    
        
    
    
        <meta property="og:image" content="https://woovictory.github.io/assets/images/Victory.jpeg"/>
    
    
    
    
    <!--STYLES-->
    <link rel="stylesheet" href="/assets/css/all.css">
    <link rel="stylesheet" href="/assets/css/jquery.fancybox.css">
    <link rel="stylesheet" href="/assets/css/thumbs.css">
    <link rel="stylesheet" href="/assets/css/tranquilpeak.css">
    <!--STYLES END-->
    
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-125410253-1"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());

        gtag('config', 'UA-125410253-1');
    </script>


    

    
        
    
</head>

    <body>
        <div id="blog">
            <!-- Define author's picture -->


    
        
            
        
    

<header id="header" data-behavior="2">
    <i id="btn-open-sidebar" class="fa fa-lg fa-bars"></i>
    <div class="header-title">
        <a
            class="header-title-link"
            href="/ "
            aria-label=""
        >
            VictoryWoo
        </a>
    </div>
    
        
            <a
                class="header-right-picture "
                href="#about"
                aria-label="Open the link: /#about"
            >
        
        
            <img class="header-picture" src="/assets/images/Victory.jpeg" alt="Author&#39;s picture"/>
        
        </a>
    
</header>

            <!-- Define author's picture -->



        
    

<nav id="sidebar" data-behavior="2">
    <div class="sidebar-container">
        
            <div class="sidebar-profile">
                <a
                    href="/#about"
                    aria-label="Read more about the author"
                >
                    <img class="sidebar-profile-picture" src="/assets/images/Victory.jpeg" alt="Author&#39;s picture"/>
                </a>
                <h4 class="sidebar-profile-name">VictoryWoo</h4>
                
                    <h5 class="sidebar-profile-bio"><p>기록을 통해 사람들과 공유하는 것을 좋아합니다.</p>
</h5>
                
            </div>
        
        
            <ul class="sidebar-buttons">
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link "
                            href="/"
                            
                            rel="noopener"
                            title="Home"
                        >
                        <i class="sidebar-button-icon fa fa-home" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">Home</span>
                    </a>
            </li>
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link "
                            href="/all-categories"
                            
                            rel="noopener"
                            title="Categories"
                        >
                        <i class="sidebar-button-icon fa fa-bookmark" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">Categories</span>
                    </a>
            </li>
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link "
                            href="/all-tags"
                            
                            rel="noopener"
                            title="Tags"
                        >
                        <i class="sidebar-button-icon fa fa-tags" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">Tags</span>
                    </a>
            </li>
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link open-algolia-search"
                            href="#search"
                            
                            rel="noopener"
                            title="Search"
                        >
                        <i class="sidebar-button-icon fa fa-search" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">Search</span>
                    </a>
            </li>
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link "
                            href="#about"
                            
                            rel="noopener"
                            title="About"
                        >
                        <i class="sidebar-button-icon fa fa-question" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">About</span>
                    </a>
            </li>
            
        </ul>
        
            <ul class="sidebar-buttons">
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link "
                            href="https://github.com/WooVictory"
                            
                                target="_blank"
                            
                            rel="noopener"
                            title="GitHub"
                        >
                        <i class="sidebar-button-icon fab fa-github" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">GitHub</span>
                    </a>
            </li>
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link "
                            href="https://www.facebook.com/WooVictory"
                            
                                target="_blank"
                            
                            rel="noopener"
                            title="Facebook"
                        >
                        <i class="sidebar-button-icon fab fa-facebook" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">Facebook</span>
                    </a>
            </li>
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link "
                            href="https://www.linkedin.com/profile/"
                            
                                target="_blank"
                            
                            rel="noopener"
                            title="LinkedIn"
                        >
                        <i class="sidebar-button-icon fab fa-linkedin" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">LinkedIn</span>
                    </a>
            </li>
            
                <li class="sidebar-button">
                    
                        <a
                            class="sidebar-button-link "
                            href="/jhsw0375@gmail.com"
                            
                            rel="noopener"
                            title="Mail"
                        >
                        <i class="sidebar-button-icon fa fa-envelope" aria-hidden="true"></i>
                        <span class="sidebar-button-desc">Mail</span>
                    </a>
            </li>
            
        </ul>
        
    </div>
</nav>

            
            <div id="main" data-behavior="2"
                 class="
                        hasCoverMetaIn
                        ">
                
<article class="post">
    
    
        <div class="post-header main-content-wrap text-left">
    
        <h1 class="post-title">
            [Java] 14. 예외 처리
        </h1>
    
    
        <div class="post-meta">
    <time datetime="2018-09-02T11:02:54+09:00">
	
		    Sep 02, 2018
    	
    </time>
    
        <span>in </span>
        
    <a class="category-link" href="/categories/Java/">Java</a>


    
</div>

    
</div>

    
    <div class="post-content markdown">
        <div class="main-content-wrap">
            <h1 id="오류error와-예외exception">오류(error)와 예외(exception)</h1>
<hr>
<p>자바 프로그램을 작성할 때 자바 문법에 맞지 않게 코드를 작성하고 컴파일하려고 하면 자바 컴파일러는 문법 오류(syntax error)를 발생시킵니다. 또한, 자바 문법에는 맞게 작성되었다 하더라도 프로그램이 실행되면서 예상하지 못한 오류가 발생할 수 있습니다.<br>
이렇게 컴퓨터 시스템이 동작하는 도중에 예상하지 못한 사태가 발생하여 실행 중인 프로그램이 영향을 받는 것을 오류(error)와 예외(exception) 두 가지로 구분할 수 있습니다.</p>
<blockquote>
<p>코드를 작성하는 과정에서 예기치 못한 수많은 에러들이 발생합니다. 예를 들면 다음과 같습니다.<br>
<strong>컴파일 전에 알게 되는 에러도 있고, 작성할 때는 아무 문제 없다가 실행시키고 나서야 발생하는 에러들이 있습니다. 이것들을 제어할 수 있어야 좀 더 안전한 프로그램을 설계할 수 있습니다.</strong></p>
</blockquote>
<p><code>에러(error)? 예외(exception)?</code></p>
<p>컴파일 에러 - 컴파일 시에 발생하는 에러<br>
런타임 에러 - 실행 시에 발생하는 에러<br>
논리적 에러 - 실행은 되지만, 의도와 다르게 동작하는 에러</p>
<p>그럼 위에서 말하는 컴파일, 런타임은 무엇이면 컴파일 에러는 무엇일까요?? 공부를 하다보니 들어본 적은 있지만, 정확하게 무엇이라고 설명하기에는 부족한 부분들이 있어서 아래에 정리해봤습니다.</p>
<ul>
<li>컴파일 에러 : 주로 문법상의 오류가 많습니다. (빨간 줄이 나타나는 에러와 같은 경우입니다. 대부분의 IDE들이 빨간줄 표시로 잡아줍니다.) 컴파일 에러가 발생하는 경우는 주로 오타, 중괄호의 짝이 맞지 않는 경우, 중복된 변수명 등의 원인이 있을 수 있습니다.</li>
<li><code>RunTime</code> : 어떤 프로그램이 실행되는 동안의 Time을 의미하며, runtime error는 어떤 프로그램이 실행되는 동안 발생하는 에러입니다.</li>
<li>런타임 에러 : 문법상으로는 에러가 없지만 실행시켰을 때 에러가 나는 경우, 대표적인 경우가 <strong>NullpointerException</strong></li>
<li><code>Compile Time</code> : 고급 언어(프로그래밍 언어)를 기계어로 변경하는 과정이라고 하며, compile error는 고급 언어(프로그래밍 언어)를 기계어로 변경하는 과정에서 발생하는 에러입니다.</li>
</ul>
<p>즉, 컴파일 에러가 발생하지 않더라도 런타임 에러가 발생할 수 있습니다.</p>
<p><img src="/img/compile_runtime.png" alt=""></p>
<h2 id="오류error">오류(error)</h2>
<p>**오류(error)**는 시스템 레벨에서 프로그램에 심각한 문제를 야기하여 실행 중인 프로그램을 종료시킵니다. 이러한 오류는 개발자가 미리 예측하여 처리할 수 없는 것이 대부분이므로, 오류에 대한 처리는 할 수 없습니다.<br>
(ex. 메모리 부족, 스택 오버플로우와 같이 발생하면 복구할 수 없는.)</p>
<h2 id="예외exception">예외(exception)</h2>
<p>**예외(exception)**는 오류와 마찬가지로 실행 중인 프로그램을 비정상적으로 종료시키지만, 발생할 수 있는 상황을 미리 예측하여 처리할 수 있습니다. 따라서 개발자는 예외 처리(Exception handling)를 통해 예외 상황을 처리할 수 있도록 코드의 흐름을 바꿀 필요가 있습니다.<br>
(발생하더라도 수습될 수 있는, 비교적 덜 심각한. 그렇기에 프로그램의 비정상적인 종료를 사전에 예방할 수 있음.)</p>
<h1 id="예외처리exception-handling">예외처리(exception handling)</h1>
<hr>
<p>자바에서 예외 처리는 이 <strong>Exception을 Handling하는 것</strong>을 말합니다. 프로그램을 제작하는 과정에서 발생할 예외에 대해 미리 코드를 작성하는 것입니다.</p>
<p>if-else문으로 처리를 해야만 했던, C에서는 예외 처리를 위한 if문과 일반 if문을 구분하기 힘들었습니다. 이 문제를 해결하기 위해 <strong>Java</strong>에서는 예외 처리에 특화된 문법을 제공합니다.</p>
<p>그것이 바로 <strong>try-catch문입니다.</strong> 사용하는 방법과 코드에 대한 설명은 아래와 같습니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span>&#123;</span><br><span class="line"><span class="comment">// 예외를 발생시킬 위험을 갖고 있는 코드(로직)</span></span><br><span class="line"><span class="comment">// 이 로직은 예외 상황과 관련있는 문장들도 고려해야 합니다.</span></span><br><span class="line">&#125;<span class="keyword">catch</span>(Exception e1)&#123;</span><br><span class="line"><span class="comment">// 예외가 발생했을 때 실행되는 로직</span></span><br><span class="line"><span class="comment">// Exception클래스는 모든 예외의 상위 클래스</span></span><br><span class="line">&#125;<span class="keyword">catch</span>(e2)&#123;</span><br><span class="line"></span><br><span class="line">&#125;</span><br><span class="line">...</span><br><span class="line"></span><br><span class="line"><span class="keyword">finally</span>&#123;</span><br><span class="line"><span class="comment">// 예외 발생 여부와 상관없이 무조건 실행될 코드</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>위의 코드에서 보면 예외를 발생시킬 위험을 갖고 있는 코드인 경우 <strong>try{ }</strong> 블록 안에 넣어둡니다. 그리고 <strong>catch</strong>를 통해 오류 메시지를 출력합니다. try안에 있는 코드를 실행하다가 오류가 발생하면 catch 구문으로 <strong>Jump</strong>합니다.</p>
<blockquote>
<p>Jump : 오류가 발생한 구문 다음의 코드는 실행되지 않는다.</p>
</blockquote>
<p>그러면 catch 메소드가 실행되게 되는데, 이 때 Exception이라는 클래스를 e로 인스턴스화하여 catch에 매개변수로 전달합니다. 그리고 Exception 이라는 클래스에 있던 getMessage라는 메소드를 호출하거나 다른 구문을 실행합니다.</p>
<p>다음은 예외가 발생했을 때 실행되는 로직의 예입니다.<br>
Sysetm.out.println(e.getMessage()); -&gt; 가장 간단한 예외 상황을 출력<br>
Sysetm.out.println(e.toString()); -&gt; 예외 상황에 대한 좀 더 자세한 정보를 출력<br>
e.printStackTrace(); -&gt; 예외 상황이 발생한 소스코드의 위치까지 출력</p>
<p>그리고 위의 코드처럼 catch 구문을 여러 개 두어서 다중 catch도 가능합니다.</p>
<p>catch 블록과 finally 블록은 선택적인 옵션으로 반드시 사용할 필요는 없습니다. 따라서 사용할 수 있는 모든 적합한 try 구문은 다음과 같습니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="number">1</span>. <span class="keyword">try</span> / <span class="keyword">catch</span></span><br><span class="line"><span class="number">2</span>. <span class="keyword">try</span> / <span class="keyword">finally</span></span><br><span class="line"><span class="number">3</span>. <span class="keyword">try</span> / <span class="keyword">catch</span> / ... / <span class="keyword">finally</span></span><br></pre></td></tr></table></figure>
<blockquote>
<p>다른 제어문과 달리 예외 처리문은 중괄호를 생략할 수 없습니다.</p>
</blockquote>
<h2 id="예외처리-메커니즘">예외처리 메커니즘</h2>
<p>자바에서 예외 처리는 아래의 그림과 순서로 진행됩니다.</p>
<ol>
<li>try 블록에 도달한 프로그램의 제어는 try 블록 내의 코드를 실행합니다.<br>
이 때 만약 예외가 발생(throw)하지 않고, finally 블록이 존재하면 프로그램의 제어는 바로 finally 블록으로 이동합니다.</li>
<li>try 블로에서 예외가 발생하면 catch 구문으로 <strong>Jump</strong>하게 되고, catch 핸들러는 다음과 같은 순서로 적절한 catch 블록을 찾게 됩니다.
<ul>
<li>스택에서 try 블록과 가장 가까운 catch 블록부터 차례대로 검사합니다.</li>
<li>만약 적절한 catch 블록을 찾지 못하면, 바로 다음 바깥쪽 try 블록 다음에 위치한 catch 블록을 차례대로 검사합니다.</li>
<li>이러한 과정을 가장 바깥쪽의 try 블록까지 계속 검사하게 됩니다.</li>
<li>그래도 적절한 catch 블록을 찾지 못하면, 예외는 처리되지 못합니다.</li>
</ul>
</li>
<li>만약 적절한 catch 블록을 찾게 되면, throw 문의 피연산자는 예외 객체의 형식 매개변수로 전달됩니다.</li>
<li>모든 예외 처리가 끝나면 프로그램의 제어는 finally 블록으로 이동합니다.(finally 블록이 있다면)</li>
<li>finally 블록이 모두 처리되면, 프로그램의 제어는 예외 처리문 바로 다음으로 이동합니다.</li>
</ol>
<img src="/img/try_catch.png" width="500" height="200">
<p>만약 1번 try 블록에서 예외가 발생하지 않고, 바깥쪽 try 블록에서도 예외가 발생하지 않으면, 6번 finally 블록이 바로 실행될 것입니다.</p>
<p>하지만, 1번 try 블록에서 예외가 발생하면 2번과 3번 catch 블록에서 해당 예외를 처리할 수 있는지 검사하게 됩니다. 만약 적절한 catch 블록을 처리하지 못하면, 바깥쪽 4,5번 catch 블록도 차례대로 검사하게 됩니다.<br>
이 때 해당 예외를 처리할 수 있는 catch 블록을 찾게 되면, 해당 catch 블록을 실행한 후 6번의 finally 블록을 실행합니다. 하지만 모든 catch 블록이 해당 예외를 처리할 수 없으면, 예외는 처리되지 못한채 해당 프로그램을 강제 종료될 것입니다.</p>
<h1 id="exception-클래스">Exception 클래스</h1>
<hr>
<p>자바에서 모든 예외의 조상 클래스가 되는 Exception 클래스는 크게 다음과 같이 구분할 수 있습니다.</p>
<ol>
<li>RuntimeException 클래스</li>
<li>그 외의 Exception 클래스의 자식 클래스</li>
</ol>
<img src="/img/exception.png" width="500" height="200">
<p>RuntimeException 클래스를 상속받는 자식 클래스들은 주로 치명적인 예외 상황을 발생시키지 않는 예외들로 구성됩니다. 따라서 try / catch 문을 사용하기보다는 프로그램을 작성하면서 예외가 발생하지 않도록 주의를 기울이는 편이 좋습니다.</p>
<p>하지만 그 외의 Exception 클래스에 속하는 자식 클래스들은 치명적인 예외 상황을 발생시키므로, 반드시 try / catch문을 사용하여 예외를 처리해야만 합니다. 따라서 자바 컴파일러는 RuntimeException 클래스 이외의 Exception 클래스의 자식 클래스에 속하는 예외가 발생할 가능성이 있는 구문에는 반드시 예외를 처리하도록 강제하고 있습니다. 만약 이러한 예외가 발생할 가능성이 있는 구문을 처리하지 않았을 때는 컴파일 시 오류를 발생시킵니다.</p>
<p>다음 예제는 PrintStream 클래스의 write() 메소드를 사용하여 byte 타입 배열의 모든 요소를 출력하는 예제입니다. 하지만 write() 메소드에서 발생할 수 있는 IOException에 대한 예외를 처리하지 않았으므로, 컴파일 시 오류가 발생합니다.</p>
<p>따라서 try / catch문을 사용하여 IOException에 대한 예외 처리까지 해주어야만 컴파일 할 수 있습니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> java.io.IOException;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">ExceptionExam</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> <span class="keyword">throws</span> IOException </span>&#123;</span><br><span class="line">        <span class="keyword">byte</span>[] list = &#123;<span class="number">1</span>,<span class="number">3</span>,<span class="number">2</span>&#125;;</span><br><span class="line">        System.out.write(list);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<h2 id="예외-처리의-계층-관계">예외 처리의 계층 관계</h2>
<p>자바에서는 예외가 발생하면, try 블록과 가장 가까운 catch 블록부터 순서대로 검사합니다. 따라서 여러 개의 catch 블록을 사용할 때는 Exception 클래스의 계층 관계에도 주의를 기울여야만 합니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line"></span><br><span class="line">    System.out.write(list);</span><br><span class="line"></span><br><span class="line">&#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line"></span><br><span class="line">    e.printStackTrace();</span><br><span class="line"></span><br><span class="line">&#125; <span class="keyword">catch</span> (IOException e) &#123;</span><br><span class="line"></span><br><span class="line">    e.printStackTrace();</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>위의 예제에서 IOException이 발생하면, 자바는 첫 번째 catch 블록부터 순서대로 해당 예외를 처리할 수 있는지를 검사합니다. 그런데 IOException은 Exception 클래스의 자식 클래스이므로, 첫 번째 catch 블록에서도 IOException을 처리할 수 있습니다.<br>
따라서 IOException을 비롯한 Exception 클래스의 자식 클래스에 해당하는 예외가 발생하면, 언제나 첫 번째 catch 블록에서만 처리될 것입니다.<br>
즉, catch 블록의 순서를 위의 코드처럼 작성하면, 두 번째 catch 블록은 영원히 실행되지 않을 것입니다.</p>
<p>따라서, IOException만을 따로 처리하고자 한다면, 다음과 같이 catch 블록의 순서를 변경해야 합니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line"></span><br><span class="line">    System.out.write(list);</span><br><span class="line"></span><br><span class="line">&#125; <span class="keyword">catch</span> (IOException e) &#123;</span><br><span class="line"></span><br><span class="line">    e.printStackTrace();</span><br><span class="line"></span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line"></span><br><span class="line">    e.printStackTrace();</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>이렇게 변경하면 IOException이 발생하면, 첫 번째 catch 블록에서 해당 예외를 처리할 것입니다. 또한, IOException 이외의 Exception 클래스의 자식 클래스에 해당하는 예외가 발생하면 두 번째 catch 블록에서 처리될 것입니다. 이처럼 범위가 더 좁은 예외를 처리하는 catch 블록을 먼저 명시하고, 범위가 더 넓은 예외를 처리하는 catch 블록은 나중에 명시해야만 정상적으로 해당 예외를 처리할 수 있습니다.</p>
<h2 id="여러-예외-타입의-동시-처리">여러 예외 타입의 동시 처리</h2>
<p><strong>Java SE 7 부터는 ‘|’ 기호를 사용하여 하나의 catch 블록에서 여러 타입의 예외를 동시에 처리할 수 있습니다.</strong></p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">this</span>.db.commit();</span><br><span class="line"></span><br><span class="line">&#125; <span class="keyword">catch</span> (IOException e) &#123;</span><br><span class="line"></span><br><span class="line">    e.printStackTrace();</span><br><span class="line"></span><br><span class="line">&#125; <span class="keyword">catch</span> (SQLException e) &#123;</span><br><span class="line"></span><br><span class="line">    e.printStackTrace();</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>위와 같이 일반적으로 작성하던 catch 블록을 아래와 같이 동시에 처리할 수 있습니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">this</span>.db.commit();</span><br><span class="line"></span><br><span class="line">&#125; <span class="keyword">catch</span> (IOException | SQLException e) &#123;</span><br><span class="line"></span><br><span class="line">    e.printStackTrace();</span><br><span class="line"></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>하지만, 둘 이상의 예외 타입을 동시에 처리하는 catch 블록에서 매개변수로 전달받은 예외 객체는 묵시적으로 final 제어자를 가지게 됩니다. 따라서 catch 블록 내에서 해당 매개변수에는 어떠한 값도 대입할 수 없습니다.(상수화가 되었기 때문)</p>
<h2 id="throwable-클래스">Throwable 클래스</h2>
<p>자바에서 Throwable 클래스는 모든 예외의 조상이 되는 Exception 클래스와 모든 오류의 조상이 되는 Error 클래스의 부모 클래스입니다. Throwable 타입과 이 클래스를 상속받은 서브 타입만이 자바 가상 머신(JVM)이나 throw 키워드에 의해 던져질 수 있습니다.</p>
<p>이 클래스에는 예외나 오류에 관한 다양한 정보를 확인할 수 있는 다음과 같은 메소드가 포함되어 있습니다.</p>
<ul>
<li>String getMessage() : 해당 throwable 객체에 대한 자세한 내용을 문자열로 반환함</li>
<li>vodi printStackTree() : 해당 throwable 객체와 표준 오류 스트림(standard error stream)에서 해당 객체의 스택 트레이스를 출력함</li>
<li>String toString() : 해당 throwable 객체에 대한 간략한 내용을 문자열로 반환함</li>
</ul>
<p>다음 예제는 일부로 숫자를 0으로 나눠서 ArithmeticException 오류를 발생시키는 예제입니다. 이렇게 발생한 오류에 대해서 Throwable 메소드를 사용하여 발생한 오류에 대한 정보를 출력합니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">package</span> com.company;</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> java.io.IOException;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">ExceptionExam</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span></span>&#123;</span><br><span class="line"></span><br><span class="line">        <span class="keyword">try</span>&#123;</span><br><span class="line">            System.out.println(<span class="number">5</span>/<span class="number">0</span>);</span><br><span class="line">        &#125;<span class="keyword">catch</span> (ArithmeticException e)&#123;</span><br><span class="line">            e.printStackTrace();</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// 결과</span></span><br><span class="line">현재 발생한 예외의 정보 : / by zero</span><br></pre></td></tr></table></figure>
<h2 id="자주-사용되는-예외-클래스">자주 사용되는 예외 클래스</h2>
<p>자바에서 자주 사용되는 예외 클래스는 다음의 표와 같습니다.</p>
<img src="/img/exception_class.png" width="500" height="200">
<h1 id="예외-발생-및-회피">예외 발생 및 회피</h1>
<hr>
<p>자바에서는 throw 키워드를 사용하여 강제로 예외를 발생시킬 수 있습니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Exception e = <span class="keyword">new</span> Exeption(<span class="string">"메시지"</span>);</span><br><span class="line"></span><br><span class="line"><span class="keyword">throw</span> e;</span><br></pre></td></tr></table></figure>
<h2 id="예외-회피하기">예외 회피하기</h2>
<p>예외 처리에 대한 부분에서 가장 헷갈리는 부분이 <strong>throw와 throws</strong>입니다.</p>
<ul>
<li><strong>throw</strong>
<ul>
<li>예외를 발생시키는 것이고, 예외를 던지는 것입니다.</li>
<li>메소드 내에서 상위 블록으로 예외를 던지는 것</li>
<li>억지로 에러를 발생시킬 때도 사용되지만 현재 메소드의 에러를 처리한 후에 상위 메소드에 에러 정보를 줌으로써 상위 메소드에서도 에러가 발생한 것을 감지할 수 있습니다.</li>
<li>실제로 exception을 throw할 때 사용하는 키워드</li>
<li><strong>throw는 강제로 예외를 발생시키는 것</strong></li>
</ul>
</li>
</ul>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">ExceptionExam</span> </span>&#123;</span><br><span class="line">    <span class="function"><span class="keyword">static</span> <span class="keyword">void</span> <span class="title">handlingException</span><span class="params">()</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> Exception();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">"호출된 메소드에서 예외가 처리됨!"</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            handlingException();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">"main() 메소드에서 예외가 처리됨!"</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// 결과</span></span><br><span class="line">호출된 메소드에서 예외가 처리됨!</span><br></pre></td></tr></table></figure>
<p>Main 함수에서 handlingException() 함수를 호출하고 그 함수 안에서는 <strong>throw</strong> 키워드를 사용하여 예외를 발생시키고 있습니다. 이 예외를 발생시키는 문장을 try / catch로 감싸고 있기 때문에 발생한 예외는 catch 블록으로 <strong>Jump</strong>하게 되고 함수 내에서 예외를 처리하게 됩니다.</p>
<ul>
<li><strong>throws</strong>
<ul>
<li>메소드나 생성자를 수행할 때 발생하는 Exception을 선언할 때 사용하는 키워드</li>
<li>예외를 던지는 것</li>
<li>현재 메소드에서 상위 메소드로 예외를 던집니다.</li>
<li>다시 말해, <strong>에외를 자신이 직접 처리하지 않고, 자신을 호출한 메소드에세 책임을 전가하는 것입니다.</strong></li>
</ul>
</li>
</ul>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="class"><span class="keyword">class</span> <span class="title">Exception04</span> </span>&#123;</span><br><span class="line"></span><br><span class="line">    <span class="function"><span class="keyword">static</span> <span class="keyword">void</span> <span class="title">handlingException</span><span class="params">()</span> <span class="keyword">throws</span> Exception </span>&#123; </span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> Exception(); </span><br><span class="line">    &#125;</span><br><span class="line">    <span class="function"><span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title">main</span><span class="params">(String[] args)</span> </span>&#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            handlingException();</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">"main() 메소드에서 예외가 처리됨!"</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">// 결과</span></span><br><span class="line">main() 메소드에서 예외가 처리됨!</span><br></pre></td></tr></table></figure>
<p>위의 코드는 <strong>throws</strong>를 사용하여 handlingException() 함수에서 발생한 예외를 전가시키는 예제입니다. 메소드를 정의할 때 <strong>throws</strong> 키워드를 추가하여 이 메소드를 호출하는 곳에서 예외 처리를 하도록 전가시킬 수 있습니다.</p>
<p><strong>handlingException() throws Exception</strong> 라는 문장을 해석해보자면, handlingException()라는 함수가 Exception 예외를 던진다는 뜻이므로 handlingException() 함수를 사용하는 곳(혹은 호출하는 곳)으로 예외를 던집니다.</p>
<p>여기서는 Main 함수에서 handlingException() 함수를 호출하였고 handlingException() 함수는 이 함수를 호출한 곳으로 예외를 던지기 때문에 Main 함수에서 예외를 처리해주어야 합니다. 하지만, 예외가 발생할 수 있기 때문에 앞에서 배운 에외 처리를 통해서 함수를 호출하는 문장을 try 블록으로 감싸줍니다. 그러면 여기서 예외가 발생하면 아래의 catch 블록으로 빠지게 되고, 결국에는 Main 함수에서 예외를 처리하게 됩니다.</p>
<h2 id="사용자-정의-예외-클래스">사용자 정의 예외 클래스</h2>
<p>자바에서는 Exception 클래스를 상속받아 자신만의 새로운 예외 클래스를 정의하여 사용할 수 있습니다.<br>
사용자 정의 예외 클래스에는 생성자뿐 아니라 필드 및 메소드도 원하는 만큼 추가할 수 있습니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">MyException</span> <span class="keyword">extends</span> <span class="title">RuntimeException</span></span>&#123;</span><br><span class="line">    MyException(String errMsg)&#123;</span><br><span class="line">        <span class="keyword">super</span>(errMsg);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>요즘에는 위와 같이 Exception 클래스가 아닌 예외 처리를 강제하지 않는 RuntimeException 클래스를 상속받아 작성하는 경우가 많습니다.</p>
<h2 id="try-with-resources문">try-with-resources문</h2>
<p>Java SE 7 부터는 사용자 자원을 자동으로 해제해주는 try-with-resources 문을 사용할 수 있습니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span>(파일을 열거나 자원을 할당하는 명령문)&#123;</span><br><span class="line">    ...</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>위와 같이 try 블록에 괄호 () 를 추가하여 파일을 열거나 자원을 할당하늠 명령문을 명시하면, 해당 try 블록이 끝나자마자 자동으로 파일을 닫거나 할당된 자원을 해제해줍니다.</p>
<p>다음은 파일에서 문자열을 한 줄 읽어오는 예제입니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">static</span> String <span class="title">readFile</span><span class="params">(String filePath)</span> <span class="keyword">throws</span> IOException </span>&#123;</span><br><span class="line">    BufferedReader br = <span class="keyword">new</span> BufferedReader(<span class="keyword">new</span> FileReader(filePath));</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="keyword">return</span> br.readLine();</span><br><span class="line">    &#125; <span class="keyword">finally</span> &#123;</span><br><span class="line">        <span class="keyword">if</span> (br != <span class="keyword">null</span>)</span><br><span class="line">            br.close();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>위와 같이 Java SE 7 이전에서는 finally 블록을 사용하여 사용한 파일을 닫아줘야 했습니다. 하지만 try-with-resources 문을 사용하면 다음과 같이 자동으로 파일의 닫기를 수행할 수 있습니다.</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">static</span> String <span class="title">readFile</span><span class="params">(String filePath)</span> <span class="keyword">throws</span> IOException </span>&#123;</span><br><span class="line">    <span class="keyword">try</span> (BufferedReader br = <span class="keyword">new</span> BufferedReader(<span class="keyword">new</span> FileReader(filePath))) &#123;</span><br><span class="line">        <span class="keyword">return</span> br.readLine();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

            


        </div>
    </div>
    <div id="post-footer" class="post-footer main-content-wrap">
        
            <div class="post-footer-tags">
                <span class="text-color-light text-small">TAGGED IN</span><br/>
                
    <a class="tag tag--primary tag--small t-link" href="/tags/Java/">Java</a>

            </div>
        
        
            <div class="post-actions-wrap">
    <nav>
        <ul class="post-actions post-action-nav">
            <li class="post-action">
                
                    
                <a
                    class="post-action-btn btn btn--default tooltip--top"
                    href="/2018/09/03/java-basic-15/"
                    data-tooltip="[Java] 15. 입력과 출력"
                    aria-label="PREVIOUS: [Java] 15. 입력과 출력"
                >
                    
                        <i class="fa fa-angle-left" aria-hidden="true"></i>
                        <span class="hide-xs hide-sm text-small icon-ml">PREVIOUS</span>
                    </a>
            </li>
            <li class="post-action">
                
                    
                <a
                    class="post-action-btn btn btn--default tooltip--top"
                    href="/2018/08/27/java-basic-13/"
                    data-tooltip="[Java] 13. 제네릭"
                    aria-label="NEXT: [Java] 13. 제네릭"
                >
                    
                        <span class="hide-xs hide-sm text-small icon-mr">NEXT</span>
                        <i class="fa fa-angle-right" aria-hidden="true"></i>
                    </a>
            </li>
        </ul>
    </nav>
    <ul class="post-actions post-action-share">
        <li class="post-action hide-lg hide-md hide-sm">
            <a
                class="post-action-btn btn btn--default btn-open-shareoptions"
                href="#btn-open-shareoptions"
                aria-label="Share this post"
            >
                <i class="fa fa-share-alt" aria-hidden="true"></i>
            </a>
        </li>
        
            
            
            <li class="post-action hide-xs">
                <a
                    class="post-action-btn btn btn--default"
                    target="new" href="https://www.facebook.com/sharer/sharer.php?u=https://woovictory.github.io/2018/09/02/java-basic-14/"
                    title="Share on Facebook"
                    aria-label="Share on Facebook"
                >
                    <i class="fab fa-facebook" aria-hidden="true"></i>
                </a>
            </li>
        
            
            
            <li class="post-action hide-xs">
                <a
                    class="post-action-btn btn btn--default"
                    target="new" href="https://twitter.com/intent/tweet?text=https://woovictory.github.io/2018/09/02/java-basic-14/"
                    title="Share on Twitter"
                    aria-label="Share on Twitter"
                >
                    <i class="fab fa-twitter" aria-hidden="true"></i>
                </a>
            </li>
        
            
            
            <li class="post-action hide-xs">
                <a
                    class="post-action-btn btn btn--default"
                    target="new" href="https://plus.google.com/share?url=https://woovictory.github.io/2018/09/02/java-basic-14/"
                    title="Share on Google+"
                    aria-label="Share on Google+"
                >
                    <i class="fab fa-google-plus" aria-hidden="true"></i>
                </a>
            </li>
        
        
            
                <li class="post-action">
                    <a
                        class="post-action-btn btn btn--default"
                        href="#disqus_thread"
                        aria-label="Leave a comment"
                    >
                        <i class="fa fa-comment"></i>
                    </a>
                </li>
            
        
        <li class="post-action">
            
                <a class="post-action-btn btn btn--default" href="#" aria-label="Back to top">
            
                <i class="fa fa-list" aria-hidden="true"></i>
            </a>
        </li>
    </ul>
</div>


        
        
            
                <div id="disqus_thread">
    <noscript>Please enable JavaScript to view the <a href="//disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</div>
            
        
    </div>
</article>



                <footer id="footer" class="main-content-wrap">
    <span class="copyrights">
        Copyrights &copy; 2020 VictoryWoo. All Rights Reserved.
    </span>
</footer>

            </div>
            
                <div id="bottom-bar" class="post-bottom-bar" data-behavior="2">
                    <div class="post-actions-wrap">
    <nav>
        <ul class="post-actions post-action-nav">
            <li class="post-action">
                
                    
                <a
                    class="post-action-btn btn btn--default tooltip--top"
                    href="/2018/09/03/java-basic-15/"
                    data-tooltip="[Java] 15. 입력과 출력"
                    aria-label="PREVIOUS: [Java] 15. 입력과 출력"
                >
                    
                        <i class="fa fa-angle-left" aria-hidden="true"></i>
                        <span class="hide-xs hide-sm text-small icon-ml">PREVIOUS</span>
                    </a>
            </li>
            <li class="post-action">
                
                    
                <a
                    class="post-action-btn btn btn--default tooltip--top"
                    href="/2018/08/27/java-basic-13/"
                    data-tooltip="[Java] 13. 제네릭"
                    aria-label="NEXT: [Java] 13. 제네릭"
                >
                    
                        <span class="hide-xs hide-sm text-small icon-mr">NEXT</span>
                        <i class="fa fa-angle-right" aria-hidden="true"></i>
                    </a>
            </li>
        </ul>
    </nav>
    <ul class="post-actions post-action-share">
        <li class="post-action hide-lg hide-md hide-sm">
            <a
                class="post-action-btn btn btn--default btn-open-shareoptions"
                href="#btn-open-shareoptions"
                aria-label="Share this post"
            >
                <i class="fa fa-share-alt" aria-hidden="true"></i>
            </a>
        </li>
        
            
            
            <li class="post-action hide-xs">
                <a
                    class="post-action-btn btn btn--default"
                    target="new" href="https://www.facebook.com/sharer/sharer.php?u=https://woovictory.github.io/2018/09/02/java-basic-14/"
                    title="Share on Facebook"
                    aria-label="Share on Facebook"
                >
                    <i class="fab fa-facebook" aria-hidden="true"></i>
                </a>
            </li>
        
            
            
            <li class="post-action hide-xs">
                <a
                    class="post-action-btn btn btn--default"
                    target="new" href="https://twitter.com/intent/tweet?text=https://woovictory.github.io/2018/09/02/java-basic-14/"
                    title="Share on Twitter"
                    aria-label="Share on Twitter"
                >
                    <i class="fab fa-twitter" aria-hidden="true"></i>
                </a>
            </li>
        
            
            
            <li class="post-action hide-xs">
                <a
                    class="post-action-btn btn btn--default"
                    target="new" href="https://plus.google.com/share?url=https://woovictory.github.io/2018/09/02/java-basic-14/"
                    title="Share on Google+"
                    aria-label="Share on Google+"
                >
                    <i class="fab fa-google-plus" aria-hidden="true"></i>
                </a>
            </li>
        
        
            
                <li class="post-action">
                    <a
                        class="post-action-btn btn btn--default"
                        href="#disqus_thread"
                        aria-label="Leave a comment"
                    >
                        <i class="fa fa-comment"></i>
                    </a>
                </li>
            
        
        <li class="post-action">
            
                <a class="post-action-btn btn btn--default" href="#" aria-label="Back to top">
            
                <i class="fa fa-list" aria-hidden="true"></i>
            </a>
        </li>
    </ul>
</div>


                </div>
                
    <div id="share-options-bar" class="share-options-bar" data-behavior="2">
        <i id="btn-close-shareoptions" class="fa fa-times"></i>
        <ul class="share-options">
            
                
                
                <li class="share-option">
                    <a
                        class="share-option-btn"
                        target="new"
                        href="https://www.facebook.com/sharer/sharer.php?u=https://woovictory.github.io/2018/09/02/java-basic-14/"
                        aria-label="Share on Facebook"
                    >
                        <i class="fab fa-facebook" aria-hidden="true"></i><span>Share on Facebook</span>
                    </a>
                </li>
            
                
                
                <li class="share-option">
                    <a
                        class="share-option-btn"
                        target="new"
                        href="https://twitter.com/intent/tweet?text=https://woovictory.github.io/2018/09/02/java-basic-14/"
                        aria-label="Share on Twitter"
                    >
                        <i class="fab fa-twitter" aria-hidden="true"></i><span>Share on Twitter</span>
                    </a>
                </li>
            
                
                
                <li class="share-option">
                    <a
                        class="share-option-btn"
                        target="new"
                        href="https://plus.google.com/share?url=https://woovictory.github.io/2018/09/02/java-basic-14/"
                        aria-label="Share on Google+"
                    >
                        <i class="fab fa-google-plus" aria-hidden="true"></i><span>Share on Google+</span>
                    </a>
                </li>
            
        </ul>
    </div>


            
        </div>
        


    
        
    

<div id="about">
    <div id="about-card">
        <div id="about-btn-close">
            <i class="fa fa-times"></i>
        </div>
        
            <img id="about-card-picture" src="/assets/images/Victory.jpeg" alt="Author&#39;s picture"/>
        
            <h4 id="about-card-name">VictoryWoo</h4>
        
            <div id="about-card-bio"><p>기록을 통해 사람들과 공유하는 것을 좋아합니다.</p>
</div>
        
        
            <div id="about-card-job">
                <i class="fa fa-briefcase"></i>
                <br/>
                <p>Android Developer</p>

            </div>
        
        
    </div>
</div>

        
        
<div id="cover" style="background-image:url('/assets/images/cover-v1.2.0.jpg');"></div>
        <!--SCRIPTS-->
<script src="/assets/js/jquery.js"></script>
<script src="/assets/js/jquery.fancybox.js"></script>
<script src="/assets/js/thumbs.js"></script>
<script src="/assets/js/tranquilpeak.js"></script>
<!--SCRIPTS END-->


    
        <script>
          var disqus_config = function() {
            this.page.url = 'https://woovictory.github.io/2018/09/02/java-basic-14/';
              
            this.page.identifier = '2018/09/02/java-basic-14/';
              
          };
          (function() {
            var d = document, s = d.createElement('script');
            var disqus_shortname = 'https-woovictory-github-io';
            s.src = '//' + disqus_shortname + '.disqus.com/embed.js';

            s.setAttribute('data-timestamp', +new Date());
            (d.head || d.body).appendChild(s);
          })();
        </script>
    




    </body>
</html>
